add evaluateRule function to find and evaluate rules by name

shahid-rasheed-stc opened the pull request.

shahid-rasheed-stc committed:
add evaluateRule function to find and evaluate rules by name

santalvarez reviewed (changes requested):
I like it but I think iterating through the entire list of rules until we find the one specified kind of defeats the purpouse of this function. Maybe it would be better to store the rules in a dictionary `[RuleName: Rule]` and then access the rule directly by doing `self.rules[ruleName]`.

Now in the evaluate function you should do self.rules.values to be able to iterate the dict.

shahid-rasheed-stc committed:
Optimize rule lookups with dictionary and add duplicate name validation

shahid-rasheed-stc commented:
Thanks for the suggestion! I switched to using a dictionary (rulesDict) for O(1) lookups, as recommended. To prevent crashes from duplicate rule names, I added a validation step in initializeRules that throws a meaningful error (RuleEngineError.duplicateRuleName) if duplicates are found.

Let me know if you have further thoughts! 😊

santalvarez commented:
Hey! Im sorry about the delay, the thing is that looking at the PR I realized we were sorting the rules on every evaluate call which is really inefficient so I made some changes to fix that. That means that we won't be able to put the rules in a dictionary as it can't be sorted. What I propose is we keep a separate dictionary of [RuleName: RuleArrayIndex] and then we can do something like this:

```swift
public init(rules: [Rule], customOperators: [Operator.Type] = []) throws {
    self.ruleDecoder = try RuleDecoder(customOperators)
    self.rules = rules
    self.rules.sort { $0.priority > $1.priority }
    self.ruleIndexByName = Dictionary(uniqueKeysWithValues:
        rules.enumerated().map { ($1.name, $0) }
    )
}
```

I like the check for duplicates you added.
